polaris/dataset/_dataset.py

```diff
@@ -170,30 +165,90 @@ def _serialize_adapters(self, value: List[Adapter]):
 
         1. Is not sensitive to the ordering of the columns or rows in the table.
         2. Purposefully does not include the meta-data (source, description, name, annotations).
-        3. For any pointer column, it uses a hash of the path instead of the file contents.
-            This is a limitation, but probably a reasonable assumption that helps practicality.
-            A big downside is that as the dataset is saved elsewhere, the hash changes.
+        3. Includes a hash for the Zarr archive.
         """
         hash_fn = md5()
 
         # Sort the columns s.t. the checksum is not sensitive to the column-ordering
-        df = table.copy(deep=True)
+        df = self.table.copy(deep=True)
         df = df[sorted(df.columns.tolist())]
 
         # Use the sum of the row-wise hashes s.t. the hash is insensitive to the row-ordering
         table_hash = pd.util.hash_pandas_object(df, index=False).sum()
         hash_fn.update(table_hash)
 
+        # If the Zarr archive exists, we hash its contents too.
+        zarr_md5sum_manifest = None
+        if self.zarr_root_path is not None:
+            zarr_hash, zarr_md5sum_manifest = compute_zarr_checksum(self.zarr_root_path)
+            hash_fn.update(zarr_hash.encode())
+
         checksum = hash_fn.hexdigest()
-        return checksum
+        return checksum, zarr_md5sum_manifest
+
+    def verify_checksum(self, md5sum: Optional[str] = None):
+        """
+        Recomputes the checksum and verifies whether it matches the stored checksum.
+
+        Warning: Slow operation
+            This operation can be slow for large datasets.
+
+        Info: Only works for locally stored datasets
+            The checksum verification only works for datasets that are stored locally in its entirety.
+            We don't have to verify the checksum for datasets stored on the Hub, as the Hub will do this on upload.
+            And if you're streaming the data from the Hub, we will check the checksum of each chunk on download.
+        """
+        if md5sum is None:
+            md5sum = self._md5sum
+        if md5sum is None:
+            logger.warning(
+                "No checksum to verify against. Specify either the md5sum parameter or "
+                "store the checksum in the dataset.md5sum attribute."
+            )
+            return
+
+        # Temporarily reset
+        # Calling self.md5sum will recompute the checksum and set it again
+        self._md5sum = None
+        if self.md5sum != md5sum:
+            raise PolarisChecksumError(
+                f"The specified checksum {md5sum} does not match the computed checksum {self.md5sum}"
+            )
+
+    @computed_field
+    @property
+    def md5sum(self) -> str:
+        """Lazily compute the checksum once needed."""
+        if not self.has_md5sum:
+            self._md5sum, self._zarr_md5sum_manifest = self._compute_checksum()
+        return self._md5sum
+
+    @md5sum.setter
+    def md5sum(self, value: str):
```

Comment:
Can any of these implementations be consolidated into some higher-up class or util module and shared? Just the ones that are identical across artifacts, to avoid them getting out of sync in the future.

Reply:
Good idea! I introduced a Mixin class!